Graphics/Slicer/Machine/Infill.hs

```diff
@@ -71,30 +71,26 @@ infillLineSegInside contour childContours line
           filterTooShort [a] = [a]
           filterTooShort (a:b:xs) = if roundToFifth (distance a b) == 0 then filterTooShort xs else a:filterTooShort (b:xs)
           getLineSegIntersections :: LineSeg -> Contour -> [Point2]
-          getLineSegIntersections myline c = saneIntersections $ lineIntersection myline <$> linesOfContour c
+          getLineSegIntersections myline c = saneIntersections $ intersectsWith (Left myline) . Left <$> linesOfContour c
             where
               linesOfContour (PointSequence contourPoints) = makeLineSegsLooped $ (\(Point2 (x,y)) -> Point2 (roundToFifth x, roundToFifth y)) <$> contourPoints
-              saneIntersections :: [Intersection] -> [Point2]
-              saneIntersections xs = concat $ mapMaybe saneIntersection xs
+              saneIntersections :: [Either Intersection PIntersection] -> [Point2]
+              saneIntersections xs = catMaybes $ mapWithNeighbors saneIntersection xs
                 where
-                  saneIntersection :: Intersection -> Maybe [Point2]
+                  saneIntersection :: Either Intersection PIntersection -> Either Intersection PIntersection -> Either Intersection PIntersection -> Maybe Point2
```

Comment:
What do these arguments refer to? Why did this type change so dramatically? It feels smelly to have this many eithers in one type.

Reply:
This function is four `where` blocks deep. Consider moving it to a top level function. A good rule of thumb is "if you needed it here, you're probably going to need it somewhere else too." Aim to keep your function definitions <20 lines --- including their where blocks.

Reply:
Each either is a different intersection between a different line, and a line segment. I switched to using neighbor checking, so i could write intersection handlers that were neighbor aware.